extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java

```diff
@@ -297,11 +265,31 @@ private StatusResponseHolder submitRequest(Request request, boolean retry)
         } else if (responseCode == 400) { // don't bother retrying if it's a bad request
           throw new IAE("Received 400 Bad Request with body: %s", response.getContent());
         } else {
-          throw new IOException(String.format("Received status [%d] with: %s", responseCode, response.getContent()));
+          throw new IOException(String.format("Received status [%d]", responseCode));
         }
       }
       catch (IOException | ChannelException e) {
-        final Duration delay = retryPolicy.getAndIncrementRetryDelay();
+
+        // Since workers are free to move tasks around to different ports, there is a chance that a task may have been
+        // moved but our view of its location has not been updated yet from ZK. To detect this case, we send a header
+        // identifying our expected recipient in the request; if this doesn't correspond to the worker we messaged, the
+        // worker will return an HTTP 404 with its ID in the response header. If we get a mismatching task ID, then
+        // we will wait for a short period then retry the request indefinitely, expecting the task's location to
+        // eventually be updated.
+
+        final Duration delay;
+        if (response != null && response.getStatus().equals(HttpResponseStatus.NOT_FOUND)) {
+          String headerId = response.getResponse().headers().get("X-Druid-Task-Id");
+          if (headerId != null && !headerId.equals(id)) {
+            log.warn("Expected worker to have taskId [%s] but has taskId [%s], will retry in 5s", id, headerId);
+            delay = Duration.standardSeconds(5);
```

Comment:
The 5 seconds was fairly arbitrary; I figured it would probably be enough time for RTR to get updated by a ZK announcement, and also that this case would happen so rarely that it wasn't worth setting up another exponential backoff mechanism for it. I'll put it as a constant.